tests/lib/rules/no-shadow-route-definition.js

```diff
@@ -0,0 +1,594 @@
+    'this.route("blog", { path: "" });',
+    'this.route("blog", { path: "/" });',
+    'this.route("blog", { path: "blog-posts" });',
+    'this.route("blog", { path: "blog-posts" }, function() {});',
+    'this.route("blog", { path: "/blog-posts" });',
+    'this.route("blog", { path: "blog-posts", otherOption: true });',
+    `this.route("blog", { path: "/" }, function() {
+      this.route("sub-blog", { path: "/" });
+    });`,
+    `this.route("blog", function() {
+      this.route("post");
+    });`,
+    `this.route("blog", { path: "/" }, function() {
+      this.route("post");
+    });`,
+    `this.route("blog", { path: "/" }, function() {
+      this.route("post", { path: "hero" });
+    });`,
+    `this.route("blog", function() {
+      this.route("post", { path: "hero" });
+    });`,
+    `this.route("blog", { path: "/blog/post" }, function() {
+      this.route("post", { path: "/" });
+    });
+    this.route("post")`,
+    `this.route("blog", function() {
+      this.route("post", { path: "/" });
+    });
+    this.route("post")`,
+
+    // With dynamic segment:
+    'this.route("blog", { path: ":blog" });',
+    'this.route("blog", { path: "/:blog" });',
+    'this.route("blog", { path: "blog/:blog_id" });',
+
+    // With dynamic segment, nested:
+    `this.route("blog", { path: "blog/:blog_id" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+    `this.route("blog", { path: "/:blog" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+    `this.route("blog", { path: ":blog" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+
+    // With wildcard segment:
+    'this.route("blog", { path: "*blog" });',
+    'this.route("blog", { path: "/*blog" });',
+    'this.route("blog", { path: "blog/*blog" });',
+
+    // With wildcard segment, nested:
+    `this.route("blog", { path: "*blog" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+    `this.route("blog", { path: "/*blog" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+    `this.route("blog", { path: "blog/*blog" }, function() {
+      this.route("post");
+    });
+    this.route("post");`,
+
+    // Not Ember's route function:
+    'test();',
+    "test('blog');",
+    "test('blog', { path: 'blog' });",
+    "test('blog', { path: '/blog' });",
+    "this.test('blog');",
+    "this.test('blog', { path: 'blog' });",
+    "this.test('blog', { path: '/blog' });",
+    "MyClass.route('blog');",
+    "MyClass.route('blog', { path: 'blog' });",
+    "MyClass.route('blog', { path: '/blog' });",
+    "route.unrelatedFunction('blog', { path: 'blog' });",
+    "this.route.unrelatedFunction('blog', { path: 'blog' });",
+  ],
+  invalid: [
+    {
+      code: `
+        this.route("blog");
+        this.route("blog");
+      `,
+      output: null,
+      errors: [
+        {
+          message: buildErrorMessage({
```

Comment:
Could we check the string error message in at least one test? This `buildErrorMessage` is pretty complicated and we actually need to test that that generates the right string output too.

Reply:
Test added.